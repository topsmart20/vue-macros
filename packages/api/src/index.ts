export * from '@vue-macros/common'

export * from './vue'
export * from './resolve'
export * from './ts'
export * from './utils'
