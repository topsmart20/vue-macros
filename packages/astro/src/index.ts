import { type AstroIntegration, type ViteUserConfig } from 'astro'
import { type Options, resolveOptions } from 'unplugin-vue-macros'
import { type Plugin } from 'vite'
import VueMacros from 'unplugin-vue-macros/vite'

export type VueMacrosOptions = Options

function findPluginAndRemove(
  name: string,
  plugins: ViteUserConfig['plugins']
): Plugin | undefined {
  const idx = plugins!.findIndex(
    (plugin) => plugin && 'name' in plugin && plugin.name === name
  )
  if (idx === -1) return
  const plugin = plugins![idx]
  plugins!.splice(idx, 1)
  return plugin as any
}

export default function (options?: VueMacrosOptions): AstroIntegration {
  return {
    name: '@vue-macros/astro',
    hooks: {
      'astro:config:setup': ({ config }) => {
        const resolvedOptions = resolveOptions(options ?? {})
        const vue = findPluginAndRemove('vite:vue', config.vite.plugins)
        const vueJsx = findPluginAndRemove('vite:vue-jsx', config.vite.plugins)

        config.vite.plugins?.push(
          VueMacros({
            ...resolvedOptions,
            plugins: {
              vue,
              vueJsx,
            },
          })
        )
      },
    },
  }
}
