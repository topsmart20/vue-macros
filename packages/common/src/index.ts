export * from './vue'
export * from './ast'
export * from './constants'
export * from './unplugin'
export { default as MagicString } from 'magic-string'
