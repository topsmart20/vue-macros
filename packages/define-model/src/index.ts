import { createUnplugin } from 'unplugin'
import {
  REGEX_SETUP_SFC,
  REGEX_VUE_SFC,
  REGEX_VUE_SUB,
  createFilter,
  detectVueVersion,
  normalizePath,
} from '@vue-macros/common'
import { transformDefineModel } from './core'
import {
  emitHelperCode,
  emitHelperId,
  helperPrefix,
  useVmodelHelperCode,
  useVmodelHelperId,
} from './core/helper'
import type { BaseOptions, MarkRequired } from '@vue-macros/common'
import type { UnpluginContextMeta } from 'unplugin'

export interface Options extends BaseOptions {
  /**
   * Unified mode, only works for Vue 2
   *
   * Converts `modelValue` to `value`
   */
  unified?: boolean
}

export type OptionsResolved = MarkRequired<
  Options,
  'include' | 'version' | 'unified'
>

function resolveOption(
  options: Options,
  framework: UnpluginContextMeta['framework']
): OptionsResolved {
  const version = options.version || detectVueVersion()
  return {
    include: [REGEX_VUE_SFC, REGEX_SETUP_SFC].concat(
      version === 2 && framework === 'webpack' ? REGEX_VUE_SUB : []
    ),
    unified: true,
    ...options,
    version,
  }
}

const name = 'unplugin-vue-define-model'

export default createUnplugin<Options | undefined, false>(
  (userOptions = {}, { framework }) => {
    const options = resolveOption(userOptions, framework)
    const filter = createFilter(options)

    return {
      name,
      enforce: 'pre',

      resolveId(id) {
        if (id.startsWith(helperPrefix)) return id
      },

      loadInclude(id) {
        return id.startsWith(helperPrefix)
      },

      load(_id) {
        const id = normalizePath(_id)
        if (id === emitHelperId) return emitHelperCode
        else if (id === useVmodelHelperId) return useVmodelHelperCode
      },

      transformInclude(id) {
        return filter(id)
      },

      transform(code, id) {
        return transformDefineModel(code, id, options.version, options.unified)
      },
    }
  }
)
