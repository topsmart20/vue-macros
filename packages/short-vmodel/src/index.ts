import { createSimpleExpression, processExpression } from '@vue/compiler-core'
import type {
  ComponentNode,
  NodeTransform,
  PlainElementNode,
  SlotOutletNode,
  TemplateNode,
  TransformContext,
} from '@vue/compiler-core'

export type Prefix = '::' | '$' | '*'
export interface Options {
  /**
   * Support :: only currently.
   *
   * @default '::'
   */
  prefix?: Prefix
}

type NodeElement =
  | PlainElementNode
  | ComponentNode
  | SlotOutletNode
  | TemplateNode

export const transformShortVmodel = ({
  prefix = '::',
}: Options = {}): NodeTransform => {
  return (node, context) => {
    if (node.type !== 1 /* NodeTypes.ELEMENT */) return
    if (prefix === '::') processDirective(node)
    else processAttribute(prefix, node, context)
  }
}

const processDirective = (node: NodeElement) => {
  for (const [i, prop] of node.props.entries()) {
    if (
      !(
        prop.type === 7 /* NodeTypes.DIRECTIVE */ &&
        prop.arg?.type === 4 /* NodeTypes.SIMPLE_EXPRESSION */ &&
        prop.arg.content.startsWith(':')
      )
    )
      continue

    const argName = prop.arg.content.slice(1)
    node.props[i] = {
      ...prop,
      name: 'model',
      arg:
        argName.length > 0
          ? { ...prop.arg, content: prop.arg.content.slice(1) }
          : undefined,
    }
  }
}

export const processAttribute = (
  prefix: string,
  node: NodeElement,
  context: TransformContext
) => {
  for (const [i, prop] of node.props.entries()) {
    if (
      !(
        prop.type === 6 /* NodeTypes.ATTRIBUTE */ &&
        prop.name.startsWith(prefix) &&
        prop.value
      )
    )
      continue

    const expLoc = prop.value.loc
    {
      // remove "
      expLoc.start.offset++
      expLoc.start.column++
      expLoc.end.offset--
      expLoc.end.column--
      expLoc.source = expLoc.source.slice(1, -1)
    }
    const simpleExpression = createSimpleExpression(
      prop.value.content,
      false,
      expLoc,
      0 /* ConstantTypes.NOT_CONSTANT */
    )
    const exp = processExpression(simpleExpression, context)

    const argName = prop.name.slice(prefix.length)
    const arg =
      argName.length > 0
        ? {
            type: 4 /* NodeTypes.SIMPLE_EXPRESSION */,
            content: argName,
            constType: 3 /* ConstantTypes.CAN_STRINGIFY */,
            isStatic: true,
            loc: {
              source: argName,
              start: {
                offset: prop.loc.start.offset + prefix.length,
                column: prop.loc.start.line + prefix.length,
                line: prop.loc.start.line,
              },
              end: {
                offset: prop.loc.start.offset + prefix.length + argName.length,
                column: prop.loc.start.line + prefix.length + argName.length,
                line: prop.loc.start.line,
              },
            },
          }
        : undefined

    node.props[i] = {
      type: 7 /* NodeTypes.DIRECTIVE */,
      name: 'model',
      arg,
      exp,
      modifiers: [],
      loc: prop.loc,
    }
  }
}
