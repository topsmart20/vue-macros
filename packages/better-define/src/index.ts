import { createUnplugin } from 'unplugin'
import {
  REGEX_SETUP_SFC,
  REGEX_VUE_SFC,
  REGEX_VUE_SUB,
  createFilter,
  detectVueVersion,
} from '@vue-macros/common'
import { RollupResolve, setResolveTSFileIdImpl } from '@vue-macros/api'
import { transformBetterDefine } from './core'
import type { BaseOptions, MarkRequired } from '@vue-macros/common'
import type { UnpluginContextMeta } from 'unplugin'
import type { PluginContext } from 'rollup'

export interface Options extends BaseOptions {
  isProduction?: boolean
}

export type OptionsResolved = MarkRequired<
  Options,
  'include' | 'version' | 'isProduction'
>

function resolveOptions(
  options: Options,
  framework: UnpluginContextMeta['framework']
): OptionsResolved {
  const version = options.version || detectVueVersion()

  return {
    include: [REGEX_VUE_SFC, REGEX_SETUP_SFC].concat(
      version === 2 && framework === 'webpack' ? REGEX_VUE_SUB : []
    ),
    isProduction: process.env.NODE_ENV === 'production',
    ...options,
    version,
  }
}

const name = 'unplugin-vue-better-define'

export default createUnplugin<Options | undefined, false>(
  (userOptions = {}, { framework }) => {
    const options = resolveOptions(userOptions, framework)
    const filter = createFilter(options)

    const { resolve, handleHotUpdate } = RollupResolve()

    return {
      name,
      enforce: 'pre',

      buildStart() {
        if (framework === 'rollup' || framework === 'vite') {
          setResolveTSFileIdImpl(resolve(this as PluginContext))
        }
      },

      transformInclude(id) {
        return filter(id)
      },

      async transform(code, id) {
        try {
          return await transformBetterDefine(code, id, options.isProduction)
        } catch (err: unknown) {
          this.warn(`${name} ${err}`)
          console.warn(err)
        }
      },

      vite: {
        configResolved(config) {
          options.isProduction = config.isProduction
        },

        handleHotUpdate,
      },
    }
  }
)
