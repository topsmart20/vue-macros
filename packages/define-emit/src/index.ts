import { createUnplugin } from 'unplugin'
import {
  type BaseOptions,
  type MarkRequired,
  REGEX_SETUP_SFC,
  REGEX_VUE_SFC,
  REGEX_VUE_SUB,
  createFilter,
  detectVueVersion,
} from '@vue-macros/common'
import { RollupResolve, setResolveTSFileIdImpl } from '@vue-macros/api'
import { generatePluginName } from '#macros' assert { type: 'macro' }
import { transformDefineEmit } from './core'
import type { PluginContext } from 'rollup'

export interface Options extends BaseOptions {
  isProduction?: boolean
}

export type OptionsResolved = MarkRequired<
  Options,
  'include' | 'version' | 'isProduction'
>

function resolveOption(options: Options): OptionsResolved {
  const version = options.version || detectVueVersion()
  return {
    include: [REGEX_VUE_SFC, REGEX_SETUP_SFC, REGEX_VUE_SUB],
    isProduction: process.env.NODE_ENV === 'production',
    ...options,
    version,
  }
}

const name = generatePluginName()

export default createUnplugin<Options | undefined, false>(
  (userOptions = {}, { framework }) => {
    const options = resolveOption(userOptions)
    const filter = createFilter(options)
    const { resolve, handleHotUpdate } = RollupResolve()

    return {
      name,
      enforce: 'pre',

      buildStart() {
        if (framework === 'rollup' || framework === 'vite') {
          setResolveTSFileIdImpl(resolve(this as PluginContext))
        }
      },

      transformInclude(id) {
        return filter(id)
      },

      transform(code, id) {
        return transformDefineEmit(code, id)
      },

      vite: {
        configResolved(config) {
          options.isProduction = config.isProduction
        },

        handleHotUpdate,
      },
    }
  },
)
