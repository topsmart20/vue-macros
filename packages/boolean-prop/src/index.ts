import { type Plugin } from 'vite'
import { type VuePluginApi, getVuePluginApi } from '@vue-macros/common'
import { generatePluginName } from '#macros' assert { type: 'macro' }
import { type Options, transformBooleanProp } from './core/index'

// legacy export
export * from './api'

const name = generatePluginName()

function rollup(options: Options = {}): Plugin {
  let api: VuePluginApi | null | undefined

  return {
    name,
    configResolved(config) {
      try {
        api = getVuePluginApi(config.plugins)
      } catch {}
    },
    buildStart(rollupOpts) {
      if (api === undefined)
        try {
          api = getVuePluginApi(rollupOpts.plugins)
        } catch (error: any) {
          this.warn(error)
          return
        }

      if (!api) return

      api.options.template ||= {}
      api.options.template.compilerOptions ||= {}
      api.options.template.compilerOptions.nodeTransforms ||= []

      api.options.template.compilerOptions.nodeTransforms.push(
        transformBooleanProp(options)
      )
    },
  }
}

export default {
  rollup,
  vite: rollup,
}
