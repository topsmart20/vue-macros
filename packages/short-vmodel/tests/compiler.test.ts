import { compileTemplate } from '@vue/compiler-sfc'
import { describe, expect, test } from 'vitest'
import { type Prefix, transformShortVmodel } from '../src'

function compile(code: string, prefix?: Prefix) {
  return compileTemplate({
    source: code,
    filename: 'anonymous.vue',
    id: 'xxx',
    compilerOptions: {
      nodeTransforms: prefix ? [transformShortVmodel({ prefix })] : undefined,
    },
  })
}

describe('compiler', () => {
  describe('double bind', () => {
    test('default name', () => {
      const original = compile(`<input v-model="foo" />`)
      const sugar = compile(`<input ::="foo" />`, '::')
      expect(sugar.code).toBe(original.code)
      expect((sugar.ast as any).children[0].props[0]).matchSnapshot()
    })

    test('w/ name', () => {
      const original = compile(`<input v-model:foo="foo" />`)
      const sugar = compile(`<input ::foo="foo" />`, '::')
      expect(sugar.code).toBe(original.code)
      expect((sugar.ast as any).children[0].props[0]).matchSnapshot()
    })
  })

  describe('dollar sign', () => {
    test('default name', () => {
      const original = compile(`<input v-model="foo" />`)
      const sugar = compile(`<input $="foo" />`, '$')
      expect(sugar.code).toBe(original.code)
      expect((sugar.ast as any).children[0].props[0]).matchSnapshot()
    })

    test('w/ name', () => {
      const original = compile(`<input v-model:foo="foo" />`)
      const sugar = compile(`<input $foo="foo" />`, '$')
      expect(sugar.code).toBe(original.code)
      expect((sugar.ast as any).children[0].props[0]).matchSnapshot()
    })
  })

  describe('asterisk sign', () => {
    test('default name', () => {
      const original = compile(`<input v-model="foo" />`)
      const sugar = compile(`<input *="foo" />`, '*')
      expect(sugar.code).toBe(original.code)
      expect((sugar.ast as any).children[0].props[0]).matchSnapshot()
    })

    test('w/ name', () => {
      const original = compile(`<input v-model:foo="foo" />`)
      const sugar = compile(`<input *foo="foo" />`, '*')
      expect(sugar.code).toBe(original.code)
      expect((sugar.ast as any).children[0].props[0]).matchSnapshot()
    })
  })
})
