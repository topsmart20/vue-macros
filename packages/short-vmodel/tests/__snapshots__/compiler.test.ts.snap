// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`compiler > asterisk sign > default name 1`] = `
{
  "arg": undefined,
  "exp": {
    "constType": 0,
    "content": "_ctx.foo",
    "isStatic": false,
    "loc": {
      "end": {
        "column": 14,
        "line": 1,
        "offset": 13,
      },
      "source": "foo",
      "start": {
        "column": 11,
        "line": 1,
        "offset": 10,
      },
    },
    "type": 4,
  },
  "loc": {
    "end": {
      "column": 15,
      "line": 1,
      "offset": 14,
    },
    "source": "*=\\"foo\\"",
    "start": {
      "column": 8,
      "line": 1,
      "offset": 7,
    },
  },
  "modifiers": [],
  "name": "model",
  "type": 7,
}
`;

exports[`compiler > asterisk sign > w/ name 1`] = `
{
  "arg": {
    "constType": 3,
    "content": "foo",
    "isStatic": true,
    "loc": {
      "end": {
        "column": 8,
        "line": 1,
        "offset": 11,
      },
      "source": "foo",
      "start": {
        "column": 8,
        "line": 1,
        "offset": 8,
      },
    },
    "type": 4,
  },
  "exp": {
    "constType": 0,
    "content": "_ctx.foo",
    "isStatic": false,
    "loc": {
      "end": {
        "column": 17,
        "line": 1,
        "offset": 16,
      },
      "source": "foo",
      "start": {
        "column": 14,
        "line": 1,
        "offset": 13,
      },
    },
    "type": 4,
  },
  "loc": {
    "end": {
      "column": 18,
      "line": 1,
      "offset": 17,
    },
    "source": "*foo=\\"foo\\"",
    "start": {
      "column": 8,
      "line": 1,
      "offset": 7,
    },
  },
  "modifiers": [],
  "name": "model",
  "type": 7,
}
`;

exports[`compiler > dollar sign > default name 1`] = `
{
  "arg": undefined,
  "exp": {
    "constType": 0,
    "content": "_ctx.foo",
    "isStatic": false,
    "loc": {
      "end": {
        "column": 14,
        "line": 1,
        "offset": 13,
      },
      "source": "foo",
      "start": {
        "column": 11,
        "line": 1,
        "offset": 10,
      },
    },
    "type": 4,
  },
  "loc": {
    "end": {
      "column": 15,
      "line": 1,
      "offset": 14,
    },
    "source": "$=\\"foo\\"",
    "start": {
      "column": 8,
      "line": 1,
      "offset": 7,
    },
  },
  "modifiers": [],
  "name": "model",
  "type": 7,
}
`;

exports[`compiler > dollar sign > w/ name 1`] = `
{
  "arg": {
    "constType": 3,
    "content": "foo",
    "isStatic": true,
    "loc": {
      "end": {
        "column": 8,
        "line": 1,
        "offset": 11,
      },
      "source": "foo",
      "start": {
        "column": 8,
        "line": 1,
        "offset": 8,
      },
    },
    "type": 4,
  },
  "exp": {
    "constType": 0,
    "content": "_ctx.foo",
    "isStatic": false,
    "loc": {
      "end": {
        "column": 17,
        "line": 1,
        "offset": 16,
      },
      "source": "foo",
      "start": {
        "column": 14,
        "line": 1,
        "offset": 13,
      },
    },
    "type": 4,
  },
  "loc": {
    "end": {
      "column": 18,
      "line": 1,
      "offset": 17,
    },
    "source": "$foo=\\"foo\\"",
    "start": {
      "column": 8,
      "line": 1,
      "offset": 7,
    },
  },
  "modifiers": [],
  "name": "model",
  "type": 7,
}
`;

exports[`compiler > double bind > default name 1`] = `
{
  "arg": undefined,
  "exp": {
    "constType": 0,
    "content": "_ctx.foo",
    "isStatic": false,
    "loc": {
      "end": {
        "column": 15,
        "line": 1,
        "offset": 14,
      },
      "source": "foo",
      "start": {
        "column": 12,
        "line": 1,
        "offset": 11,
      },
    },
    "type": 4,
  },
  "loc": {
    "end": {
      "column": 16,
      "line": 1,
      "offset": 15,
    },
    "source": "::=\\"foo\\"",
    "start": {
      "column": 8,
      "line": 1,
      "offset": 7,
    },
  },
  "modifiers": [],
  "name": "model",
  "type": 7,
}
`;

exports[`compiler > double bind > w/ name 1`] = `
{
  "arg": {
    "constType": 3,
    "content": "foo",
    "isStatic": true,
    "loc": {
      "end": {
        "column": 13,
        "line": 1,
        "offset": 12,
      },
      "source": ":foo",
      "start": {
        "column": 9,
        "line": 1,
        "offset": 9,
      },
    },
    "type": 4,
  },
  "exp": {
    "constType": 0,
    "content": "_ctx.foo",
    "isStatic": false,
    "loc": {
      "end": {
        "column": 18,
        "line": 1,
        "offset": 17,
      },
      "source": "foo",
      "start": {
        "column": 15,
        "line": 1,
        "offset": 14,
      },
    },
    "type": 4,
  },
  "loc": {
    "end": {
      "column": 19,
      "line": 1,
      "offset": 18,
    },
    "source": "::foo=\\"foo\\"",
    "start": {
      "column": 8,
      "line": 1,
      "offset": 7,
    },
  },
  "modifiers": [],
  "name": "model",
  "type": 7,
}
`;
